File: job-application-assistant/backend/app/routes/alerts.py
"""Job alerts and notifications routes."""
from datetime import datetime
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.database import get_db
from app.models.alert import JobAlert, Notification
from app.models.user import User
from app.routes.auth import get_current_user
from app.services.job_search import job_search_service
from pydantic import BaseModel

router = APIRouter(prefix="/api/alerts", tags=["Alerts & Notifications"])


class JobAlertCreate(BaseModel):
    name: str
    keywords: Optional[str] = ""
    location: Optional[str] = ""
    job_type: Optional[str] = ""
    experience_level: Optional[str] = ""
    salary_min: Optional[int] = None
    industry: Optional[str] = ""
    frequency: str = "daily"

class JobAlertUpdate(BaseModel):
    name: Optional[str] = None
    keywords: Optional[str] = None
    location: Optional[str] = None
    job_type: Optional[str] = None
    experience_level: Optional[str] = None
    salary_min: Optional[int] = None
    industry: Optional[str] = None
    frequency: Optional[str] = None
    is_active: Optional[bool] = None

class JobAlertResponse(BaseModel):
    id: int
    name: str
    keywords: Optional[str] = ""
    location: Optional[str] = ""
    job_type: Optional[str] = ""
    experience_level: Optional[str] = ""
    salary_min: Optional[int] = None
    industry: Optional[str] = ""
    frequency: str
    is_active: bool
    last_triggered: Optional[datetime] = None
    created_at: datetime
    matching_jobs_count: Optional[int] = 0

    class Config:
        from_attributes = True

class NotificationResponse(BaseModel):
    id: int
    title: str
    message: str
    notification_type: str
    reference_type: Optional[str] = None
    reference_id: Optional[int] = None
    is_read: bool
    created_at: datetime

    class Config:
        from_attributes = True


@router.post("/", response_model=JobAlertResponse, status_code=201)
async def create_alert(
    data: JobAlertCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Create a new job alert."""
    alert = JobAlert(
        user_id=current_user.id,
        name=data.name,
        keywords=data.keywords,
        location=data.location,
        job_type=data.job_type,
        experience_level=data.experience_level,
        salary_min=data.salary_min,
        industry=data.industry,
        frequency=data.frequency,
    )
    db.add(alert)
    db.commit()
    db.refresh(alert)
    return alert


@router.get("/", response_model=list[JobAlertResponse])
async def list_alerts(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """List all job alerts."""
    alerts = db.query(JobAlert).filter(
        JobAlert.user_id == current_user.id
    ).order_by(JobAlert.created_at.desc()).all()

    results = []
    for alert in alerts:
        resp = JobAlertResponse.model_validate(alert)
        # Count matching jobs
        search_result = await job_search_service.search_jobs(
            query=alert.keywords or "",
            location=alert.location or "",
            job_type=alert.job_type or "",
            experience_level=alert.experience_level or "",
            industry=alert.industry or "",
            page=1,
            page_size=1,
        )
        resp.matching_jobs_count = search_result.get("total", 0)
        results.append(resp)
    return results


@router.put("/{alert_id}", response_model=JobAlertResponse)
async def update_alert(
    alert_id: int,
    data: JobAlertUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Update a job alert."""
    alert = db.query(JobAlert).filter(
        JobAlert.id == alert_id,
        JobAlert.user_id == current_user.id,
    ).first()
    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")

    for key, value in data.model_dump(exclude_unset=True).items():
        setattr(alert, key, value)
    alert.updated_at = datetime.utcnow()
    db.commit()
    db.refresh(alert)
    return alert


@router.delete("/{alert_id}")
async def delete_alert(
    alert_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Delete a job alert."""
    alert = db.query(JobAlert).filter(
        JobAlert.id == alert_id,
        JobAlert.user_id == current_user.id,
    ).first()
    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")

    db.delete(alert)
    db.commit()
    return {"message": "Alert deleted successfully"}


@router.post("/{alert_id}/trigger")
async def trigger_alert(
    alert_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Manually trigger a job alert to find matching jobs."""
    alert = db.query(JobAlert).filter(
        JobAlert.id == alert_id,
        JobAlert.user_id == current_user.id,
    ).first()
    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")

    search_result = await job_search_service.search_jobs(
        query=alert.keywords or "",
        location=alert.location or "",
        job_type=alert.job_type or "",
        experience_level=alert.experience_level or "",
        industry=alert.industry or "",
    )

    alert.last_triggered = datetime.utcnow()
    db.commit()

    return {
        "message": "Alert triggered",
        "matching_jobs": search_result.get("total", 0),
        "jobs": search_result.get("results", []),
    }


# --- Notifications ---
@router.get("/notifications", response_model=list[NotificationResponse])
async def list_notifications(
    unread_only: bool = False,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """List notifications for the current user."""
    query = db.query(Notification).filter(Notification.user_id == current_user.id)
    if unread_only:
        query = query.filter(Notification.is_read == False)
    return query.order_by(Notification.created_at.desc()).limit(50).all()


@router.post("/notifications/{notification_id}/read")
async def mark_notification_read(
    notification_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Mark a notification as read."""
    notif = db.query(Notification).filter(
        Notification.id == notification_id,
        Notification.user_id == current_user.id,
    ).first()
    if not notif:
        raise HTTPException(status_code=404, detail="Notification not found")

    notif.is_read = True
    db.commit()
    return {"message": "Notification marked as read"}


@router.post("/notifications/read-all")
async def mark_all_notifications_read(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Mark all notifications as read."""
    db.query(Notification).filter(
        Notification.user_id == current_user.id,
        Notification.is_read == False,
    ).update({"is_read": True})
    db.commit()
    return {"message": "All notifications marked as read"}


@router.get("/notifications/unread-count")
async def get_unread_count(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Get count of unread notifications."""
    count = db.query(Notification).filter(
        Notification.user_id == current_user.id,
        Notification.is_read == False,
    ).count()
    return {"unread_count": count}

File: job-application-assistant-node/backend/routes/alerts.js
const express = require('express');
const router = express.Router();
const { db } = require('../db');
const { authMiddleware } = require('../auth');

router.use(authMiddleware);

// POST /api/alerts
router.post('/', (req, res) => {
    try {
        const { name, keywords, location, job_type, experience_level, salary_min, industry, frequency } = req.body;
        if (!name || !keywords) return res.status(400).json({ error: 'Name and keywords are required' });

        const now = new Date().toISOString();
        const result = db.prepare(
            'INSERT INTO job_alerts (user_id, name, keywords, location, job_type, experience_level, salary_min, industry, frequency, created_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)'
        ).run(req.user.id, name, keywords, location || null, job_type || null, experience_level || null, salary_min || null, industry || null, frequency || 'daily', now);

        res.status(201).json({
            message: 'Job alert created',
            alert: { id: result.lastInsertRowid, name, keywords, frequency: frequency || 'daily', is_active: 1, created_at: now }
        });
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

// GET /api/alerts
router.get('/', (req, res) => {
    try {
        const alerts = db.prepare('SELECT * FROM job_alerts WHERE user_id = ? ORDER BY created_at DESC').all(req.user.id);
        res.json(alerts);
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

// PUT /api/alerts/:id
router.put('/:id', (req, res) => {
    try {
        const existing = db.prepare('SELECT * FROM job_alerts WHERE id = ? AND user_id = ?').get(req.params.id, req.user.id);
        if (!existing) return res.status(404).json({ error: 'Alert not found' });

        const { name, keywords, location, job_type, experience_level, salary_min, industry, frequency, is_active } = req.body;
        db.prepare(
            'UPDATE job_alerts SET name = COALESCE(?, name), keywords = COALESCE(?, keywords), location = COALESCE(?, location), job_type = COALESCE(?, job_type), experience_level = COALESCE(?, experience_level), salary_min = COALESCE(?, salary_min), industry = COALESCE(?, industry), frequency = COALESCE(?, frequency), is_active = COALESCE(?, is_active) WHERE id = ? AND user_id = ?'
        ).run(name || null, keywords || null, location || null, job_type || null, experience_level || null, salary_min || null, industry || null, frequency || null, is_active !== undefined ? (is_active ? 1 : 0) : null, req.params.id, req.user.id);

        const updated = db.prepare('SELECT * FROM job_alerts WHERE id = ?').get(req.params.id);
        res.json({ message: 'Alert updated', alert: updated });
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

// DELETE /api/alerts/:id
router.delete('/:id', (req, res) => {
    try {
        const result = db.prepare('DELETE FROM job_alerts WHERE id = ? AND user_id = ?').run(req.params.id, req.user.id);
        if (result.changes === 0) return res.status(404).json({ error: 'Alert not found' });
        res.json({ message: 'Alert deleted' });
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

// POST /api/alerts/trigger
router.post('/trigger', (req, res) => {
    try {
        const alerts = db.prepare('SELECT * FROM job_alerts WHERE user_id = ? AND is_active = 1').all(req.user.id);
        const now = new Date().toISOString();

        for (const alert of alerts) {
            db.prepare('UPDATE job_alerts SET last_triggered = ? WHERE id = ?').run(now, alert.id);
            db.prepare(
                'INSERT INTO notifications (user_id, title, message, notification_type, reference_type, reference_id, created_at) VALUES (?, ?, ?, ?, ?, ?, ?)'
            ).run(req.user.id, `New jobs for: ${alert.name}`, `Found new job matches for "${alert.keywords}"`, 'job_alert', 'job_alert', alert.id, now);
        }

        res.json({ message: `Triggered ${alerts.length} alerts`, alerts_triggered: alerts.length });
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

// GET /api/alerts/notifications
router.get('/notifications', (req, res) => {
    try {
        const notifications = db.prepare('SELECT * FROM notifications WHERE user_id = ? ORDER BY created_at DESC').all(req.user.id);
        res.json(notifications);
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

// PUT /api/alerts/notifications/:id/read
router.put('/notifications/:id/read', (req, res) => {
    try {
        db.prepare('UPDATE notifications SET is_read = 1 WHERE id = ? AND user_id = ?').run(req.params.id, req.user.id);
        res.json({ message: 'Notification marked as read' });
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

// PUT /api/alerts/notifications/read-all
router.put('/notifications/read-all', (req, res) => {
    try {
        db.prepare('UPDATE notifications SET is_read = 1 WHERE user_id = ?').run(req.user.id);
        res.json({ message: 'All notifications marked as read' });
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

// GET /api/alerts/notifications/unread-count
router.get('/notifications/unread-count', (req, res) => {
    try {
        const row = db.prepare('SELECT COUNT(*) as count FROM notifications WHERE user_id = ? AND is_read = 0').get(req.user.id);
        res.json({ unread_count: row.count });
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

module.exports = router;
